internal/pkg/client/cache/dir.go

```diff
@@ -22,75 +46,268 @@ const (
 	// for image downloads to be cached in
 	DirEnv = "SINGULARITY_CACHEDIR"
 
-	// RootDefault specifies the directory inside of ${HOME} that images are
-	// cached in by default.
-	// Uses "~/.singularity/cache" which will not clash with any 2.x cache
+	// BasedirDefault specifies the default value of the cache base directory. The
+	// path of the user's home directory is prepended at runtime. Ultimately, the
+	// default cache base directory is ~/.singularity/cache".
+	// Using "~/.singularity/cache" also does not clash with any 2.x cache
 	// directory.
-	RootDefault = ".singularity/cache"
+	BasedirDefault = ".singularity"
+
+	// RootDefault is the default cache root directory created within the base
+	// directory. This value is not supposed to be set by the user.
+	rootDefault = "cache"
+
+	// StateInitialized represents the state of a give cache after successful
+	// initialization
+	StateInitialized = "initialized"
+
+	// StateInvalid represents the state of an invalid cache
+	StateInvalid = "invalid"
 )
 
-var root string
+// SingularityCache is an opaque structure representing a cache
+type SingularityCache struct {
+	// BaseDir is the cache base directory. This directory is never entirely
+	// deleted since it can be specified by the user via the DirEnv environment
+	// variable.
+	BaseDir string
 
-// Root is the root location where all of singularity caching happens. Library, Shub,
-// and oci image formats supported by containers/image repository will be cached inside
-//
-// Defaults to ${HOME}/.singularity/cache
-func Root() string {
-	updateCacheRoot()
+	// rootDir is the cache root directory, within basedir. This is the directory
+	// Singularity actually manages, i.e., that can safely be deleted as
+	// opposed to the base directory that is potentially managed (passed in)
+	// by the user.
+	rootDir string
+
+	// State of the cache. We enable manual change of the state mainly for
+	// testing.
+	State string
+
+	// Default specifies if the handle points at the default image cache or
+	// not. This enables quick lookup.
+	Default bool
```

Comment:
If I think it is just something I forgot to remove after including a state. I will remove it. 

As BaseDir not being useful, you are referring to your expectations and I personally do not know what they are but I disagree with this: it is a key piece for testing and it makes the cache handle self-contained (`internal/pkg/client/cache/testing.go`). This is in fact the initial motivation for modifying the implementation: I will use this to revisit the testing code in `internal/app/singulairty/`, we do *not* have all the required capabilities to implement adequate tests relying on caches. BaseDir is *not* equivalent to `rootDir != ""` since the value of BaseDir is necessary.

Reply:
Comments added to the code

Reply:
Unit tests should _never_ operate against the default cache directory (as it is impossible to assert the state of that directory). That will leave a particular path of code untested, yes, but that path should be as small as:

```
basedir := os.Getenv("SINGULARITY_CACHEDIR")
if basedir == "" {
    // this branch never gets tested during unit testing
    // defaultCacheDir() itself _could_ be tested by itself
    basedir = defaultCacheDir()
}
cacheDir := filepath.Join(basedir, "cache")
```

I'm judging the lack of usefulness of BaseDir from the point of view of separation of concerns, not personal preference. That's the reason why I am asking what the idea is here.

Reply:
I would agree with you that we should never use the default cache but so far we were most of the time running tests using the default cache directory; it is exactly what I am trying to change here. But the point here is: having a base directory allows us to change the configuration of the cache to reach some error cases. Not having this will create a headache for testing. If you disagree with having the base directory there, I do not really see a path forward at the moment.
I really do not know what to do: if we remove the base directory, I am back to the initial situation when it is about reaching error caches and I do not see alternative options.

Finally, my intent was absolutely not to question your motivation. I never suggested or even considered that your comments were based on personal preferences but at the same time, it is not possible for me to understand what your expectation precisely mean, and will not take the liberty to guess. I hope I do not go through as having a negative attitude that was clearly not my intent.